File: src/10.graph/19.78.go
package graph

func Subsets(nums []int) [][]int {
	ans := [][]int{}
	onPath := []int{}
	var backtrack func(r int)

	backtrack = func(r int) {
		ans = append(ans, append([]int{}, onPath...))

		for i := r; i < len(nums); i++ {
			onPath = append(onPath, nums[i])
			backtrack(i + 1)
			onPath = onPath[:len(onPath)-1]
		}
	}

	backtrack(0)

	return ans
}

File: src/10.graph/19.78.ts
export function subsets(nums: number[]): number[][] {
    const ans: number[][] = [];
    const onPath: number[] = [];
    const n = nums.length;
    const backtrack = (r: number) => {
        ans.push([...onPath]);

        for (let i = r; i < n; i++) {
            onPath.push(nums[i]);
            backtrack(i + 1);
            onPath.pop();
        }
    };

    backtrack(0);

    return ans;
}
